File: edgequake/sdks/typescript/src/types/documents.ts
/**
 * Document types.
 *
 * WHY: Updated to match Rust documents_types.rs exactly.
 * Rust types include detailed pipeline fields (source_type, current_stage,
 * stage_progress), cost info, and structured pagination.
 *
 * @module types/documents
 * @see edgequake/crates/edgequake-api/src/handlers/documents_types.rs
 */

import type { Timestamp } from "./common.js";

// ── Upload ────────────────────────────────────────────────────

export interface UploadDocumentRequest {
  content: string;
  title?: string;
  metadata?: Record<string, unknown>;
  /** Whether to process asynchronously (default: false). */
  async_processing?: boolean;
  /** Track ID for batch grouping. If not provided, one will be generated. */
  track_id?: string;
  /** Enable gleaning (multiple extraction passes). Default: true. */
  enable_gleaning?: boolean;
  /** Maximum gleaning passes (1-3 recommended). Default: 1. */
  max_gleaning?: number;
  /** Enable LLM-powered description summarization. Default: true. */
  use_llm_summarization?: boolean;
}

export interface UploadDocumentResponse {
  document_id: string;
  status: string;
  /** Task ID (only set when async_processing is true). */
  task_id?: string;
  /** Track ID for batch grouping. */
  track_id: string;
  /** ID of existing document if this is a duplicate. */
  duplicate_of?: string;
  /** Number of chunks created (only set for sync processing). */
  chunk_count?: number;
  /** Number of entities extracted (only set for sync processing). */
  entity_count?: number;
  /** Number of relationships extracted (only set for sync processing). */
  relationship_count?: number;
  /** Cost information (only set for sync processing). */
  cost?: DocumentCostInfo;
}

/** Cost information for a processed document. */
export interface DocumentCostInfo {
  /** Total cost in USD. */
  total_cost_usd: number;
  /** Formatted cost string (e.g., "$0.0045"). */
  formatted_cost: string;
  /** Total input tokens used. */
  input_tokens: number;
  /** Total output tokens used. */
  output_tokens: number;
  /** Total tokens (input + output). */
  total_tokens: number;
  /** LLM model used. */
  llm_model?: string;
  /** Embedding model used. */
  embedding_model?: string;
}

// ── File Upload ───────────────────────────────────────────────

export interface UploadFileMetadata {
  title?: string;
  metadata?: Record<string, unknown>;
}

export interface BatchUploadResponse {
  documents: UploadDocumentResponse[];
  total: number;
  succeeded: number;
  failed: number;
}

/** Response from single file upload. */
export type UploadFileResponse = UploadDocumentResponse;

// ── List ──────────────────────────────────────────────────────

export interface ListDocumentsQuery {
  /** Page number (1-indexed). Default: 1. */
  page?: number;
  /** Page size. Default: 20. */
  page_size?: number;
  /** Filter by status. */
  status?: string;
  /** Search text in title/content. */
  search?: string;
}

/** Status counts for document filtering. */
export interface StatusCounts {
  pending: number;
  processing: number;
  completed: number;
  /** Documents with partial failure (processed but 0 entities). */
  partial_failure: number;
  failed: number;
  cancelled: number;
}

/** List documents response. */
export interface ListDocumentsResponse {
  documents: DocumentSummary[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
  has_more: boolean;
  /** Status counts for all documents (not just current page). */
  status_counts: StatusCounts;
}

/** Document summary (list item). */
export interface DocumentSummary {
  id: string;
  title?: string;
  /** Original file name. */
  file_name?: string;
  /** First 200 characters of content. */
  content_summary?: string;
  /** Total content length in characters. */
  content_length?: number;
  /** Number of chunks. */
  chunk_count: number;
  /** Number of entities extracted. */
  entity_count?: number;
  /** Processing status (legacy — use current_stage for pipelines). */
  status?: string;
  /** Error message if processing failed. */
  error_message?: string;
  /** Track ID for batch grouping. */
  track_id?: string;
  created_at?: string;
  updated_at?: string;
  /** Total cost in USD. */
  cost_usd?: number;
  /** Input tokens used. */
  input_tokens?: number;
  /** Output tokens used. */
  output_tokens?: number;
  /** Total tokens (input + output). */
  total_tokens?: number;
  /** LLM model used. */
  llm_model?: string;
  /** Embedding model used. */
  embedding_model?: string;
  /** Document source type (pdf, markdown, text). */
  source_type?: string;
  /** Current ingestion stage - see docs for stage names. */
  current_stage?: string;
  /** Progress within current stage (0.0 to 1.0). */
  stage_progress?: number;
  /** Human-readable message for current stage. */
  stage_message?: string;
  /** Linked PDF document ID (only if source_type is "pdf"). */
  pdf_id?: string;
}

/** Document detail response with full content. */
export interface DocumentDetail {
  id: string;
  title?: string;
  file_name?: string;
  /** Full document content. */
  content?: string;
  content_summary?: string;
  content_length?: number;
  chunk_count: number;
  entity_count?: number;
  status?: string;
  error_message?: string;
  track_id?: string;
  created_at?: string;
  updated_at?: string;
  cost_usd?: number;
  input_tokens?: number;
  output_tokens?: number;
  total_tokens?: number;
  llm_model?: string;
  embedding_model?: string;
  source_type?: string;
  current_stage?: string;
  stage_progress?: number;
  stage_message?: string;
  pdf_id?: string;
  metadata?: Record<string, unknown>;
}

// ── Legacy aliases ────────────────────────────────────────────
/** @deprecated Use DocumentSummary */
export type DocumentInfo = DocumentSummary;

// ── Track ─────────────────────────────────────────────────────

export interface TrackStatusResponse {
  track_id: string;
  status: string;
  progress?: number;
  documents?: Array<{
    document_id: string;
    status: string;
    error?: string;
  }>;
  created_at?: Timestamp;
  updated_at?: Timestamp;
}

// ── Scan ──────────────────────────────────────────────────────

export interface ScanDirectoryRequest {
  path: string;
  recursive?: boolean;
  max_files?: number;
  extensions?: string[];
}

export interface ScanDirectoryResponse {
  total_files: number;
  queued: number;
  skipped: number;
  track_id: string;
}

// ── Reprocess ─────────────────────────────────────────────────

export interface ReprocessRequest {
  max_reprocess?: number;
}

export interface ReprocessResponse {
  reprocessed: number;
  message: string;
}

export interface RecoverStuckRequest {
  stuck_threshold_minutes?: number;
}

export interface RecoverStuckResponse {
  recovered: number;
  message: string;
}

// ── Deletion Impact ───────────────────────────────────────────

export interface DeletionImpactResponse {
  document_id: string;
  chunks_affected: number;
  entities_affected: number;
  relationships_affected: number;
}

// ── Chunks ────────────────────────────────────────────────────

export interface RetryChunksResponse {
  retried: number;
  message: string;
}

export interface FailedChunkInfo {
  chunk_id: string;
  error: string;
  created_at: Timestamp;
}

export interface FailedChunksResponse {
  chunks: FailedChunkInfo[];
  total: number;
}

// ── PDF ───────────────────────────────────────────────────────

/**
 * Options for PDF upload (v0.4.0+).
 *
 * Vision pipeline: renders each page to an image and sends it to a
 * multimodal LLM for high-fidelity Markdown extraction.
 */
export interface PdfUploadOptions {
  title?: string;
  metadata?: Record<string, unknown>;
  track_id?: string;
  /** Enable LLM vision pipeline for high-fidelity extraction. Default: false. */
  enable_vision?: boolean;
  /** Override vision provider (e.g. "openai", "ollama"). */
  vision_provider?: string;
  /** Override vision model (e.g. "gpt-4o", "gemma3"). */
  vision_model?: string;
  /** Re-process even if document already exists. Default: false. */
  force_reindex?: boolean;
}

/** @deprecated Use PdfUploadOptions — kept for backward compatibility. */
export interface PdfUploadMetadata {
  title?: string;
  metadata?: Record<string, unknown>;
}

export interface PdfUploadResponse {
  pdf_id: string;
  document_id?: string;
  status: string;
  track_id: string;
  message?: string;
}

export interface ListPdfsQuery {
  limit?: number;
  offset?: number;
  status?: string;
}

export interface PdfInfo {
  pdf_id: string;
  document_id?: string;
  filename: string;
  status: string;
  file_size: number;
  page_count?: number;
  created_at: Timestamp;
  /** Extraction method used: "vision", "text", or "ocr" (0.4.0+). */
  extraction_method?: string;
}

export interface PdfStatusResponse extends PdfInfo {
  error_message?: string;
  markdown_content?: string;
  track_id?: string;
}

export interface PdfContentResponse {
  pdf_id: string;
  markdown: string;
  page_count: number;
}

export interface PdfProgressResponse {
  track_id: string;
  status: string;
  progress: number;
  current_page?: number;
  total_pages?: number;
  message?: string;
}

export interface PdfRetryResponse {
  pdf_id: string;
  status: string;
  message: string;
}

// ── Delete All ────────────────────────────────────────────────

export interface DeleteAllResponse {
  deleted: number;
  message: string;
}

File: edgequake/sdks/rust/src/types/documents.rs
//! Document-related types.

use serde::{Deserialize, Serialize};

use super::common::PaginationInfo;

/// Response from uploading a document.
#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct UploadDocumentResponse {
    pub id: String,
    #[serde(default)]
    pub status: Option<String>,
    #[serde(default)]
    pub message: Option<String>,
    #[serde(default)]
    pub track_id: Option<String>,
}

/// Document summary in list responses.
#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct DocumentSummary {
    pub id: String,
    #[serde(default)]
    pub file_name: Option<String>,
    #[serde(default)]
    pub title: Option<String>,
    #[serde(default)]
    pub status: Option<String>,
    #[serde(default)]
    pub file_size: Option<u64>,
    #[serde(default)]
    pub mime_type: Option<String>,
    #[serde(default)]
    pub entity_count: Option<u32>,
    #[serde(default)]
    pub chunk_count: Option<u32>,
    #[serde(default)]
    pub created_at: Option<String>,
    #[serde(default)]
    pub updated_at: Option<String>,
}

/// Response from listing documents.
#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct ListDocumentsResponse {
    #[serde(default)]
    pub documents: Vec<DocumentSummary>,
    #[serde(default)]
    pub pagination: Option<PaginationInfo>,
}

/// Response from tracking document processing status.
#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct TrackStatusResponse {
    pub track_id: String,
    pub status: String,
    #[serde(default)]
    pub progress: Option<f64>,
    #[serde(default)]
    pub message: Option<String>,
    #[serde(default)]
    pub document_id: Option<String>,
}

/// Response from directory scanning.
#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct ScanResponse {
    #[serde(default)]
    pub files_found: u32,
    #[serde(default)]
    pub files_queued: u32,
    #[serde(default)]
    pub files_skipped: u32,
}

/// Response from deletion impact analysis.
#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct DeletionImpactResponse {
    #[serde(default)]
    pub entity_count: u32,
    #[serde(default)]
    pub relationship_count: u32,
    #[serde(default)]
    pub chunk_count: u32,
}

/// Options for PDF upload (v0.4.0+).
///
/// Vision pipeline renders each page to an image and passes it to a
/// multimodal LLM for high-fidelity Markdown extraction.
#[derive(Debug, Clone, Default, Serialize)]
pub struct PdfUploadOptions {
    /// Enable LLM vision pipeline for high-fidelity extraction.
    #[serde(skip_serializing_if = "std::ops::Not::not")]
    pub enable_vision: bool,
    /// Override vision provider (e.g. "openai", "ollama").
    #[serde(skip_serializing_if = "Option::is_none")]
    pub vision_provider: Option<String>,
    /// Override vision model (e.g. "gpt-4o", "gemma3").
    #[serde(skip_serializing_if = "Option::is_none")]
    pub vision_model: Option<String>,
    /// Human-readable title.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub title: Option<String>,
    /// Batch track ID.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub track_id: Option<String>,
    /// Re-process even if document already exists.
    #[serde(skip_serializing_if = "std::ops::Not::not")]
    pub force_reindex: bool,
}

/// PDF upload response.
#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct PdfUploadResponse {
    /// Primary identifier returned by the API (v0.4.0+).
    #[serde(default)]
    pub pdf_id: Option<String>,
    /// Legacy field — same as pdf_id on older servers.
    #[serde(default)]
    pub id: Option<String>,
    #[serde(default)]
    pub document_id: Option<String>,
    #[serde(default)]
    pub status: Option<String>,
    #[serde(default)]
    pub track_id: Option<String>,
    #[serde(default)]
    pub message: Option<String>,
    #[serde(default)]
    pub estimated_time_seconds: Option<u64>,
    #[serde(default)]
    pub duplicate_of: Option<String>,
}

impl PdfUploadResponse {
    /// Return the canonical PDF ID regardless of which field the server used.
    pub fn canonical_id(&self) -> Option<&str> {
        self.pdf_id.as_deref().or_else(|| self.id.as_deref())
    }
}

/// PDF document metadata returned by list/get endpoints.
#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct PdfInfo {
    pub pdf_id: String,
    #[serde(default)]
    pub document_id: Option<String>,
    /// Original filename uploaded.
    #[serde(default, alias = "file_name")]
    pub filename: Option<String>,
    #[serde(default)]
    pub status: Option<String>,
    #[serde(default)]
    pub file_size: Option<u64>,
    #[serde(default)]
    pub page_count: Option<u32>,
    /// Extraction method: "vision", "text", or "ocr" (v0.4.0+).
    #[serde(default)]
    pub extraction_method: Option<String>,
    #[serde(default)]
    pub created_at: Option<String>,
}

/// PDF progress response.
#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct PdfProgressResponse {
    pub track_id: String,
    pub status: String,
    #[serde(default)]
    pub progress: Option<f64>,
}

/// PDF content response.
#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct PdfContentResponse {
    pub id: String,
    #[serde(default)]
    pub markdown: Option<String>,
}

/// Scan request parameters.
#[derive(Debug, Clone, Serialize)]
pub struct ScanRequest {
    pub path: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub recursive: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub extensions: Option<Vec<String>>,
}
